frame/democracy/src/lib.rs

```diff
@@ -0,0 +1,2084 @@
+		/// because the conviction period is over.
+		///
+		/// The dispatch origin of this call must be _Signed_.
+		///
+		/// - `target`: The account of the vote to be removed; this account must have voted for
+		///   referendum `index`.
+		/// - `index`: The index of referendum of the vote to be removed.
+		///
+		/// Weight: `O(R + log R)` where R is the number of referenda that `target` has voted on.
+		///   Weight is calculated for the maximum number of vote.
+		#[pallet::weight(T::WeightInfo::remove_other_vote(T::MaxVotes::get()))]
+		pub fn remove_other_vote(
+			origin: OriginFor<T>,
+			target: T::AccountId,
+			asset_id: T::AssetId,
+			index: ReferendumIndex,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let scope = if target == who { UnvoteScope::Any } else { UnvoteScope::OnlyExpired };
+			Self::try_remove_vote(&target, asset_id, index, scope)?;
+			Ok(())
+		}
+
+		/// Enact a proposal from a referendum. For now we just make the weight be the maximum.
+		#[pallet::weight(T::BlockWeights::get().max_block)]
+		pub fn enact_proposal(
+			origin: OriginFor<T>,
+			proposal_id: ProposalId<T::Hash, T::AssetId>,
+			index: ReferendumIndex,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			Self::do_enact_proposal(proposal_id, index)
+		}
+
+		/// Permanently place a proposal into the blacklist. This prevents it from ever being
+		/// proposed again.
+		///
+		/// If called on a queued public or external proposal, then this will result in it being
+		/// removed. If the `ref_index` supplied is an active referendum with the proposal hash,
+		/// then it will be cancelled.
+		///
+		/// The dispatch origin of this call must be `BlacklistOrigin`.
+		///
+		/// - `proposal_hash`: The proposal hash to blacklist permanently.
+		/// - `ref_index`: An ongoing referendum whose hash is `proposal_hash`, which will be
+		/// cancelled.
+		///
+		/// Weight: `O(p)` (though as this is an high-privilege dispatch, we assume it has a
+		///   reasonable value).
+		#[pallet::weight((T::WeightInfo::blacklist(T::MaxProposals::get()), DispatchClass::Operational))]
+		#[transactional]
+		pub fn blacklist(
+			origin: OriginFor<T>,
+			proposal_hash: T::Hash,
+			asset_id: T::AssetId,
+			maybe_ref_index: Option<ReferendumIndex>,
+		) -> DispatchResult {
+			T::BlacklistOrigin::ensure_origin(origin)?;
+
+			let id = ProposalId { hash: proposal_hash, asset_id };
+
+			// Insert the proposal into the blacklist.
+			let permanent = (T::BlockNumber::max_value(), Vec::<T::AccountId>::new());
+			Blacklist::<T>::insert(&id, permanent);
+
+			// Remove the queued proposal, if it's there.
+			PublicProps::<T>::try_mutate(|props| -> DispatchResult {
+				if let Some(index) = props.iter().position(|p| p.1 == id) {
+					let (prop_index, ..) = props.remove(index);
+					if let Some((whos, amount)) = DepositOf::<T>::take(prop_index) {
+						for who in whos.into_iter() {
+							T::NativeCurrency::transfer_held(
```

Comment:
so, we are doing best effort. is that not attack? i mean on_unbalanced allowed to have negative balance after slash. why we changed that?

Reply:
Fundamentally there is change in how the transfer is handled in MutateHold trait against slashing the reserve in the original, in our case with transfer_held() immediately transfer the amount to treasury account, and doesn't cause imbalance, best_effort is to handle the approximation. Added to report the difference, in transfer with debug_assert

Reply:
not sure about this. why not to slash who up to imbalance? right now it slashes to zero? imbalance was before. do not we support imbalnce?

Reply:
We don't need to handle imbalance, as it is taken care my fungible traits, and also orml traits doesn't support negative imbalance hence we changed.